File: Git Archaeologist/perplexity.human.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        FastScanner fs = new FastScanner(System.in);
        StringBuilder out = new StringBuilder();
        int t = fs.nextInt();

        while (t-- > 0) {
            int n = fs.nextInt();
            int q = fs.nextInt();

            long[] risk = new long[n];
            for (int i = 0; i < n; i++) risk[i] = fs.nextLong();

            ArrayList<Integer>[] children = new ArrayList[n];
            for (int i = 0; i < n; i++) children[i] = new ArrayList<>();

            for (int i = 0; i < n - 1; i++) {
                int p = fs.nextInt();
                int v = fs.nextInt();
                children[p].add(v);
            }

            int log = 1;
            while ((1 << log) <= n) log++;

            int[][] up = new int[log][n];
            long[][] mx = new long[log][n];
            int[] depth = new int[n];

            ArrayDeque<Integer> queue = new ArrayDeque<>();
            queue.add(0);
            mx[0][0] = risk[0];

            while (!queue.isEmpty()) {
                int u = queue.poll();
                for (int v : children[u]) {
                    depth[v] = depth[u] + 1;
                    up[0][v] = u;
                    mx[0][v] = Math.max(risk[v], risk[u]);
                    queue.add(v);
                }
            }

            for (int k = 1; k < log; k++) {
                for (int v = 0; v < n; v++) {
                    int mid = up[k - 1][v];
                    up[k][v] = up[k - 1][mid];
                    mx[k][v] = Math.max(mx[k - 1][v], mx[k - 1][mid]);
                }
            }

            while (q-- > 0) {
                int a = fs.nextInt();
                int b = fs.nextInt();
                int x = a, y = b;
                long best = Math.max(risk[a], risk[b]);

                if (depth[x] < depth[y]) {
                    int tmp = x;
                    x = y;
                    y = tmp;
                }

                int diff = depth[x] - depth[y];
                for (int k = 0; k < log; k++) {
                    if ((diff & (1 << k)) != 0) {
                        best = Math.max(best, mx[k][x]);
                        x = up[k][x];
                    }
                }

                if (x != y) {
                    for (int k = log - 1; k >= 0; k--) {
                        if (up[k][x] != up[k][y]) {
                            best = Math.max(best, mx[k][x]);
                            best = Math.max(best, mx[k][y]);
                            x = up[k][x];
                            y = up[k][y];
                        }
                    }

                    best = Math.max(best, risk[x]);
                    best = Math.max(best, risk[y]);
                    best = Math.max(best, risk[up[0][x]]);
                    x = up[0][x];
                }

                int distance = depth[a] + depth[b] - 2 * depth[x];
                out.append(x).append(' ').append(distance).append(' ').append(best).append('\n');
            }
        }

        System.out.print(out);
    }

    static class FastScanner {
        private final InputStream in;
        private final byte[] buffer = new byte[1 << 16];
        private int ptr, len;

        FastScanner(InputStream in) {
            this.in = in;
        }

        private int read() throws IOException {
            if (ptr >= len) {
                len = in.read(buffer);
                ptr = 0;
                if (len <= 0) return -1;
            }
            return buffer[ptr++];
        }

        long nextLong() throws IOException {
            int c;
            do c = read(); while (c <= 32);
            long x = 0;
            while (c > 32) {
                x = x * 10 + c - '0';
                c = read();
            }
            return x;
        }

        int nextInt() throws IOException {
            return (int) nextLong();
        }
    }
}

File: Git Archaeologist/perplexity.human.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;

    while (T--) {
        int n, q;
        cin >> n >> q;

        vector<long long> risk(n);
        for (auto &x : risk) cin >> x;

        vector<vector<int>> g(n);
        vector<int> parent(n, 0), depth(n, 0);

        for (int i = 0; i < n - 1; ++i) {
            int p, v;
            cin >> p >> v;
            parent[v] = p;
            g[p].push_back(v);
        }

        int LOG = 1;
        while ((1 << LOG) <= n) ++LOG;

        vector<vector<int>> up(LOG, vector<int>(n));
        vector<vector<long long>> mx(LOG, vector<long long>(n));

        queue<int> qu;
        qu.push(0);
        up[0][0] = 0;
        mx[0][0] = risk[0];

        while (!qu.empty()) {
            int u = qu.front();
            qu.pop();

            for (int v : g[u]) {
                depth[v] = depth[u] + 1;
                up[0][v] = u;
                mx[0][v] = max(risk[v], risk[u]);
                qu.push(v);
            }
        }

        for (int k = 1; k < LOG; ++k) {
            for (int v = 0; v < n; ++v) {
                int mid = up[k - 1][v];
                up[k][v] = up[k - 1][mid];
                mx[k][v] = max(mx[k - 1][v], mx[k - 1][mid]);
            }
        }

        while (q--) {
            int a, b;
            cin >> a >> b;

            long long best = max(risk[a], risk[b]);
            int x = a, y = b;

            if (depth[x] < depth[y]) swap(x, y);

            int diff = depth[x] - depth[y];
            for (int k = LOG - 1; k >= 0; --k) {
                if (diff & (1 << k)) {
                    best = max(best, mx[k][x]);
                    x = up[k][x];
                }
            }

            if (x != y) {
                for (int k = LOG - 1; k >= 0; --k) {
                    if (up[k][x] != up[k][y]) {
                        best = max(best, mx[k][x]);
                        best = max(best, mx[k][y]);
                        x = up[k][x];
                        y = up[k][y];
                    }
                }
                best = max(best, risk[x]);
                best = max(best, risk[y]);
                best = max(best, risk[up[0][x]]);
                x = up[0][x];
            }

            int lca = x;
            int dist = depth[a] + depth[b] - 2 * depth[lca];

            cout << lca << ' ' << dist << ' ' << best << '\n';
        }
    }
}
